src/irmin-pack/irmin_pack_layers.ml

```diff
@@ -790,37 +751,35 @@ struct
     end
 
     module CopyFromLower = struct
-      let on_current_upper t f =
-        let contents = X.Contents.CA.current_upper t.X.Repo.contents in
-        let nodes = X.Node.CA.current_upper t.X.Repo.node in
-        let commits = X.Commit.CA.current_upper t.X.Repo.commit in
-        f contents nodes commits
-
-      let copy_tree contents nodes t root =
-        let node k =
-          X.Node.CA.copy_from_lower ~dst:nodes t.X.Repo.node k >|= fun () ->
-          (* We have to ensure that a node is entirely flushed to disk.
-             If we rely on the implicit flush of io, it could be that
-             only a part of a node if flushed to disk when RO tries to
-             read it. *)
-          X.Node.CA.flush t.node
```

Comment:
Do you mind leaving this in? We don't have a test when a readonly instance tries to read the store while `self_contained` runs, and this flush prevented a bug in that case.

Reply:
it's flush after the full copy is complete, isn't this enough?

Reply:
The [`save`](https://github.com/mirage/irmin/blob/master/src/irmin-pack/inode.ml#L611) can call `unsafe_append` several times for a node in [`copy_from_lower`](https://github.com/mirage/irmin/blob/master/src/irmin-pack/inode_layers.ml#L162) , meaning that a flush can occur midway through a node copy and not reoccur again after. 
The RO first tries to read the upper, and then the lower, so it will see the partial node in upper and fail. 
I guess the best is to have a test to reproduce this anyway, so remove it for now if you want and I'll make an issue.

Reply:
I think it's best to have a proper test case for this. Also, #1150 will not use `save` anymore so that should be a problem anymore?